Add SNS verify_message

tlvenn opened the pull request:
Hi @benwilson512,

Here is a initial stab at the SNS verify_message, please let me know your comments and how I should proceed with testing. Do we want to use something like Exvcr with a fixture to properly test the whole flow ?

This resolves #248

Thanks!

benwilson512 commented on lib/ex_aws/sns/public_key_cache.ex:
```diff
@@ -0,0 +1,59 @@
+  end
+
+  defp fetch_certificate(cert_url) do
+    with {:ok, 200, _resp_headers, client} <- :hackney.get(cert_url, [], "", []),
```
We should use `:hackney.get(cert_url, [], "", [:with_body])` because
1) it's simpler
2) Right now we only read the body on `200`. However the body doesn't go away for other status codes, and will result in basically a memory leak if not read.

tlvenn replied:
Pretty neat, I usually don't interact with Hackney directly and did not see that option.

benwilson512 commented on lib/ex_aws/sns/public_key_cache.ex:
```diff
@@ -0,0 +1,59 @@
+    case length(pem_entries) do
+      0 -> {:error, "invalid argument"}
+      x when x > 1 -> {:error, "found multiple PEM entries, expected only 1"}
+      x when x == 1 ->
```
This could more easily be:
```
case pem_entries do
  [entry] ->
    try do # etc
  entries ->
    {:error, "Invalid PEM entries: #{inspect entries}"}
end
```

benwilson512 commented on lib/ex_aws/sns/public_key_cache.ex:
```diff
@@ -0,0 +1,59 @@
+  end  
+  
+  defp get_public_key(cert) do
+    {:ok, :public_key.der_decode(:RSAPublicKey, cert |> elem(1) |> elem(7) |> elem(2))}
```
Not sure what value there is in wrapping this with `:ok`.

tlvenn replied:
Yep fixed.

benwilson512 commented on lib/ex_aws/sns/public_key_cache.ex:
```diff
@@ -0,0 +1,59 @@
+  end
+
+  def get(certUrl) do
+      case :ets.lookup(__MODULE__, certUrl) do
```
should be `cert_url`

benwilson512 commented on lib/ex_aws/sns.ex:
```diff
@@ -169,6 +169,77 @@ defmodule ExAws.SNS do
+        |> get_string_to_sign
+        |> verify(message_params["Signature"], public_key)
+    else
+      {:error, error} -> {false, error}
```
Not really a fan of `{false, error}`. What's wrong with `{:ok, msg} | {:error, reason}` as a return signature?

On that note, we need typespecs.

benwilson512 commented on lib/ex_aws/sns.ex:
```diff
@@ -169,6 +169,77 @@ defmodule ExAws.SNS do
+  defp get_string_to_sign(message_params) do
+    message_params
+    |> Map.take(get_params_to_sign(message_params["Type"]))
+    |> Enum.reduce("", fn {key, value}, acc -> acc <> key <> "\n" <> value <> "\n" end)
```
This would probably be better built up as iodata and then concatenated at the end.

tlvenn replied:
I simplified it with `|> Enum.map_join(&(elem(&1,0)<>"\n"<>elem(&1,1)<>"\n"))`

benwilson512 replied:
Sorry I wasn't super clear. What I had in mind was something more like
```
|> Enum.map(fn {key, value} -> [to_string(key), "\n", to_string(value), "\n"] end)
|> IO.iodata_to_binary
```

It's more efficient to build up IOData in this case than concatenating binaries. Doesn't make a huge difference but it's also pretty simple code. I almost always prefer `{key, value}` over `elem` btw, just seems clearer.

tlvenn replied:
Ha thanks for enlightening me on this, I did not know about iodata_to_binary. And yes I agree, pattern matching on tuples is much nicer and cleaner than trying to use shorthand function.

benwilson512 commented on lib/ex_aws/sns.ex:
```diff
@@ -169,6 +169,77 @@ defmodule ExAws.SNS do
+    case message_type do
+      "Notification" -> {:ok, (@notification_params -- @optional_notification_params) ++ @signature_params}
+      "SubscriptionConfirmation" -> {:ok, @confirmation_params ++ @signature_params}
+      "UnsubscribeConfirmation" -> {:ok, @confirmation_params ++ @signature_params}
```
Are these types coming from amazon? Is there any reason to expect it would NOT be one of these?

tlvenn replied:
Those types are indeed coming from Amazon SNS documentation. It should always be one of those.

benwilson512 commented on lib/ex_aws/sns.ex:
```diff
@@ -169,6 +169,77 @@ defmodule ExAws.SNS do
+  @message_types ["SubscriptionConfirmation", "UnsubscribeConfirmation", "Notification"]
+
+  @doc "Verify message signature"
+  def verify_message(message_params) do
```
Do we want to automatically verify all messages?

benwilson512 commented on lib/ex_aws/sns/public_key_cache.ex:
```diff
@@ -0,0 +1,59 @@
+defmodule ExAws.SNS.PublicKeyCache do
```
Do these public keys ever expire?

tlvenn replied:
SNS certificates are valid for 1year.
<img width="524" alt="sns_amazonaws_com" src="https://cloud.githubusercontent.com/assets/3340262/20256778/95f24880-aa80-11e6-9322-0b95a963fedf.png">

tlvenn commented:
Hi @benwilson512,

Sorry for the late feedback on this issue, got side tracked pretty badly last 2 weeks...
I just pushed a new commit, hopefully addressing all your concerns.

tlvenn commented:
btw shouldn't I use http_client instead of using hackney directly ?

benwilson512 commented:
Yes yeah, we can't assume people are using hackney.

tlvenn commented:
Hi @benwilson512, the PR should be ready for final review and hopefully if all good, ready to be merged

tlvenn commented:
Hi @benwilson512 , when you get the time, could you please do final review and merge if ready ? Thanks !

benwilson512 commented:
I think this looks good, thanks!
